### ISSUE TITLE
Add investigation area, mission information as external reference in API

### ISSUE THREAD

**Issue by tloubrieu-jpl** [NASA-PDS/pds-api#37]
Get the investigation area from the xml label, rewrite how target and instrument are read, get them also from the XML label

**Comment by al-niessner** [comment]
1. use postman
2. URL: base_url/collections/limit=10
3. params??

dockerize postman for now

**Comment by tloubrieu-jpl** [comment]
Start from repo : https://github.com/NASA-PDS/registry-api-service

What we want in the products output is: https://app.swaggerhub.com/apis/PDS_APIs/pds_federated_api/0.1.dev#/product

We want to implement the serialization of investigation_area, target and instruments attributes.

**Comment by al-niessner** [comment]
Getting the error below during build. Any ideas (as in expected so just fix it)? Oh, maybe jdk version -- using JDK 11.

```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] src/main/java/gov/nasa/pds/api/engineering/controllers/MyCollectionsApiController.java:[24,8] gov.nasa.pds.api.engineering.controllers.MyCollectionsApiController is not abstract and does not override abstract method collectionsByLidvid(java.lang.String) in gov.nasa.pds.api.base.CollectionsApi
[INFO] 1 error
```

**Comment by al-niessner** [comment]
Fails the same with JDK 8

**Comment by al-niessner** [comment]
Elastic search is running but it needs to be populated. Updated pds-registry-app (pulled to master) and tried to build docker image as README states. It is failing with an error that I do not recognize (see below) but seems to indicate that what used to be a gzipped tarfile is no longer. It is also part of the MAVEN build and not directly linked to the dockerfile. I cannot proceed on this ticket until this resolved nor is clear what the appropriate action is to resolve it as in change the pom or change the release. Please advise.


```
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/commons/commons-compress/1.11/commons-compress-1.11.jar (426 kB at 202 kB/s)
Feb 17, 2021 7:19:39 PM org.apache.http.client.protocol.ResponseProcessCookies processCookies
WARNING: Invalid cookie header: "Set-Cookie: _octo=GH1.1.836262781.1613589579; Path=/; Domain=github.com; Expires=Thu, 17 Feb 2022 19:19:39 GMT; Secure; SameSite=Lax". Invalid 'expires' attribute: Thu, 17 Feb 2022 19:19:39 GMT
Feb 17, 2021 7:19:39 PM org.apache.http.client.protocol.ResponseProcessCookies processCookies
WARNING: Invalid cookie header: "Set-Cookie: logged_in=no; Path=/; Domain=github.com; Expires=Thu, 17 Feb 2022 19:19:39 GMT; HttpOnly; Secure; SameSite=Lax". Invalid 'expires' attribute: Thu, 17 Feb 2022 19:19:39 GMT
[INFO] Downloading: https://github.com/nasa-pds/pds-registry-mgr-elastic/releases/download/v4.1.0-SNAPSHOT/registry-manager-4.1.0-SNAPSHOT-bin.tar.gz
[INFO] 9/?
[INFO] downloaded 9
[INFO] Expanding: /usr/local/registry-app-fb9a90a869ef4a503201873ffd6cbbb506aa0c9c/target/registry-manager-4.1.0-SNAPSHOT-bin.tar.gz into /usr/local/registry-app-fb9a90a869ef4a503201873ffd6cbbb506aa0c9c/target
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:42 min
[INFO] Finished at: 2021-02-17T19:19:39Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.googlecode.maven-download-plugin:download-maven-plugin:1.5.0:wget (get-registry-subpackage) on project pds-registry-app: IO Error: Error while expanding /usr/local/registry-app-fb9a90a869ef4a503201873ffd6cbbb506aa0c9c/target/registry-manager-4.1.0-SNAPSHOT-bin.tar.gz: Not in GZIP format -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by al-niessner** [comment]
Looking at the [release](https://github.com/NASA-PDS/pds-registry-mgr-elastic/releases/tag/4.1.0-SNAPSHOT) now and it appears that it was not done correctly. I happened 8 days ago according to github which explains why it is now failing when it used to work previously. The file size for the tar.gz and zip are the same -- not a perfect or exacting measure but a suspicious one at the very least -- indicating that the zip file may be inadvertently been installed as the tar.gz.

**Comment by tloubrieu-jpl** [comment]
Hi @al-niessner ,

The right link to get the release file from is currently:
https://github.com/NASA-PDS/pds-registry-mgr-elastic/releases/download/4.1.0-SNAPSHOT/registry-manager-4.1.0-SNAPSHOT-bin.tar.gz

The issue is that we have a discrepency between snapshot tags and stable tags. The later start with a v (for example v0.4.1) the former does not start with a v. That is a bug in our continuous integration component. I will try to fix it.

**Comment by tloubrieu-jpl** [comment]
@al-niessner , 

There was 2 different issues (new dependency of registry-manager and harvest on a third repo and links to snapshot releases) and I think I located both but I cannot make the pds-registry-app release work without some help from @nutjob4life since it involves the roundup action.

We'll wait until he recovered from the extreme weather in texas.

**Comment by tloubrieu-jpl** [comment]
It took some time to Al to set up maven project in eclipse but everything is fine.

**Comment by al-niessner** [comment]
With that fixed, the harvest commands are no longer working.

```
$ harvest  -c /var/local/harvest/conf/examples/registry.xml -o /var/local/harvest/output
[SUMMARY] Output directory: /var/local/harvest/output
[SUMMARY] Output format: json
[SUMMARY] Reading configuration from /var/local/harvest/conf/examples/registry.xml
[WARN] Registry is not configured
[WARN] Invalid bundle directory: /path/to/bundle
```

then

```
$ registry-manager load-data -es http://es:9200 -file /var/local/harvest/output
[ERROR] Missing required parameter '-dir'
```

Will spend next workday trying to figure out what changed in registry that it now requires different arguments.

**Comment by al-niessner** [comment]
Okay, changed -file to -dir and now see this:

```
registry-manager load-data -es http://es:9200 -dir /var/local/harvest/output 
Elasticsearch URL: http://es:9200
            Index: registry

Updating schema with fields from /var/local/harvest/output/fields.txt
Updated 0 fields
Loading file: /var/local/harvest/output/registry-docs.json
Loading file: /var/local/harvest/output/refs-docs.json
```

It seems that the harvest command created 3 empty files.

**Comment by al-niessner** [comment]
Found it. 'archive' changed to 'bundle' in the harvest configuration file.

**Comment by al-niessner** [comment]
Confused. Eclipse is happy as in no red X's. It is freshly rebuilt as well. When maven does a build:

```
ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project registry-api-service: Compilation failure
[ERROR] /usr/local/registry-api-service-c937c91a6f05237dbb74863e7b1ef35c0918cbe3/src/main/java/gov/nasa/pds/api/engineering/controllers/MyCollectionsApiController.java:[24,8] gov.nasa.pds.api.engineering.controllers.MyCollectionsApiController is not abstract and does not override abstract method productsOfACollection(java.lang.String,java.lang.Integer,java.lang.Integer,java.util.List<java.lang.String>,java.util.List<java.lang.String>,java.lang.Boolean) in gov.nasa.pds.api.base.CollectionsApi
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

Hints on how to synchronize maven and eclipse so that I can use the latter to fix the former would be welcome.

**Comment by al-niessner** [comment]
Is there a table about that maps a given maven dependency to a github repo. Specifically:

```
<dependency>
	<groupId>gov.nasa.pds</groupId>
	<artifactId>api</artifactId>
	<version>0.0.1-SNAPSHOT</version>
</dependency>
```

Goes to which repo?

**Comment by al-niessner** [comment]
Back on the move. Had to delete the version that was in the cache then tell eclipse to update the maven object with a force. It now seems to have a different jar as both eclipse and command line now agree.

**Comment by al-niessner** [comment]
General question about properties and overriding them. I am finding myself having to modify registery-api-service:src/main/resources/application.properties to change elasticSearch.host. Can this be done via environment variables or some other way than editing the properties file?

**Comment by jordanpadams** [comment]
@tdddblog @tloubrieu-jpl ☝️

**Comment by al-niessner** [comment]
Just started receiving this message today. Is this my network because it is windy or did something change that I am not aware of:

```
[ERROR] Failed to execute goal on project registry-api-service: Could not resolve dependencies for project gov.nasa.pds:registry-api-service:jar:0.0.1: Could not transfer artifact org.elasticsearch:elasticsearch:jar:7.9.3 from/to central (https://repo.maven.apache.org/maven2): GET request of: org/elasticsearch/elasticsearch/7.9.3/elasticsearch-7.9.3.jar from central failed: Premature end of Content-Length delimited message body (expected: 12,778,265; received: 12,350,286) -> [Help 1]
```

**Comment by tloubrieu-jpl** [comment]
@al-niessner for the application.properties I think you can hae an additional application.properties in the target directory. That will override the values in the resources directory. You should not commit this specific application.properties of course.

For the error on the elasticsearch dependency, is it a one time thing or do have it alway since this morning ? I was able to build a minute ago.

**Comment by al-niessner** [comment]
Target directory? What is the target directory? Target of what or whom?

Just started today and has repeated several times. I am doing it for the fourth time to see if it fails with the same number of bytes. Likely my network as the wind blows tree leaves (violent swaying is more accurate) between me and the cell tower making the connection less than ideal. Yes, my only home internet is my cell hotspot.

**Comment by tloubrieu-jpl** [comment]
@al-niessner  Sorry I should have said 'target' directory. When you build with maven every output is created in a 'target' directory at the base of your project folder.
So actually you should have a 'target/classes/application.properties' deployed. I think you can override this one.

Otherwise there is this method which should also work https://stackoverflow.com/questions/23563363/how-to-override-application-properties-during-production-in-spring-boot

Ok, I thought you were joking with the wind... That is weird that it keeps failing with the number of bytes though. I've seen this issue https://issues.sonatype.org/browse/MVNCENTRAL-360 which says it is time out related...

**Comment by al-niessner** [comment]
Sorry for being such a bother, but am I supposed to update often or never? I just updated my branch and merged in registry-api-service:master to fix some conflicts. Now eclipse is complaining that these two items do not exist:

```
import gov.nasa.pds.api.engineering.elasticsearch.business.CollectionProductRefBusinessObject;
import gov.nasa.pds.api.engineering.elasticsearch.business.CollectionProductRelationships;
```

Are they supposed to be part of the registery-api-serivce repo but I did not get them somehow? I updated maven and it seemed happy enough meaning that if the missing items are in yet another jar from somewhere then my POM is not retrieving them quite right.

**Comment by tloubrieu-jpl** [comment]
Sorry @al-niessner , I missed to push those files when I merged my update on relationships.

Are you working on a branch yourself ? You should, I think that would spare you these hick ups whenever the master branch is updated.

**Comment by al-niessner** [comment]
I am on a branch. My [pull request](https://github.com/NASA-PDS/registry-api-service/pull/7/) told me there was conflict with master and I wanted some of the changes there to fix some dummy code I put in to fix changes to an api.base interface that I had no idea how to actually implement. In doing so, I discovered these missing items. So, no, working on a branch is not saving me from hiccups. Hopefully the missing items will be pushed soon.

**Comment by tloubrieu-jpl** [comment]
@al-niessner, Yes I missed to mention that I pushed the missing items. You can try again.

**Comment by al-niessner** [comment]
Thank you. I have everything updated now and build again without errors.

I am using a settings.xml to see if that will get me pass the timeout problem. I am just going to write it off as a slow internet connection. Whether or not the slowness is because of the wind or my carrier hates me does not really matter.

**Comment by al-niessner** [comment]
What is it supposed to look like when `mvn spring-boot:run` works? It appears to start for me but localhost:8080 gives me a "cannot be reached" error. localhost:9200 is there and I see it is v 7.10 of elastic search. When I scroll through the reams output from `mvn spring-boot:run` I see several stack traces but do not know if these are normal, acceptable, or oh-nos. Here is an example of one of them:

```
2021-02-25 20:39:59.911 DEBUG 59 --- [           main] o.a.catalina.core.AprLifecycleListener   : The Apache Tomcat Native library could not be found using names [tcnative-1, libtcnative-1] on the java.library.path [/usr/java/packages/lib:/usr/lib/x86_64-linux-gnu/jni:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/jni:/lib:/usr/lib]. The errors reported were [Can't load library: /usr/local/registry-api-service-689dc2cce5648f560f50ffdf4bbed1318da1acc9/bin/libtcnative-1.so, Can't load library: /usr/local/registry-api-service-689dc2cce5648f560f50ffdf4bbed1318da1acc9/bin/liblibtcnative-1.so, no tcnative-1 in java.library.path: [/usr/java/packages/lib, /usr/lib/x86_64-linux-gnu/jni, /lib/x86_64-linux-gnu, /usr/lib/x86_64-linux-gnu, /usr/lib/jni, /lib, /usr/lib], no libtcnative-1 in java.library.path: [/usr/java/packages/lib, /usr/lib/x86_64-linux-gnu/jni, /lib/x86_64-linux-gnu, /usr/lib/x86_64-linux-gnu, /usr/lib/jni, /lib, /usr/lib]]

org.apache.tomcat.jni.LibraryNotFoundError: Can't load library: /usr/local/registry-api-service-689dc2cce5648f560f50ffdf4bbed1318da1acc9/bin/libtcnative-1.so, Can't load library: /usr/local/registry-api-service-689dc2cce5648f560f50ffdf4bbed1318da1acc9/bin/liblibtcnative-1.so, no tcnative-1 in java.library.path: [/usr/java/packages/lib, /usr/lib/x86_64-linux-gnu/jni, /lib/x86_64-linux-gnu, /usr/lib/x86_64-linux-gnu, /usr/lib/jni, /lib, /usr/lib], no libtcnative-1 in java.library.path: [/usr/java/packages/lib, /usr/lib/x86_64-linux-gnu/jni, /lib/x86_64-linux-gnu, /usr/lib/x86_64-linux-gnu, /usr/lib/jni, /lib, /usr/lib]
	at org.apache.tomcat.jni.Library.<init>(Library.java:102) ~[tomcat-embed-core-9.0.36.jar:9.0.36]
<...snip...>
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) ~[spring-boot-2.3.1.RELEASE.jar:2.3.1.RELEASE]
	at gov.nasa.pds.api.engineering.SpringBootMain.main(SpringBootMain.java:27) ~[classes/:na]
```

They are not all tomcat but be related. Here is another example:

```
2021-02-25 20:40:00.708 DEBUG 59 --- [           main] o.s.v.b.OptionalValidatorFactoryBean     : Failed to set up a Bean Validation provider

javax.validation.NoProviderFoundException: Unable to create a Configuration, because no Bean Validation provider could be found. Add a provider like Hibernate Validator (RI) to your classpath.
	at javax.validation.Validation$GenericBootstrapImpl.configure(Validation.java:291) ~[validation-api-2.0.1.Final.jar:na]
	at org.springframework.validation.beanvalidation.LocalValidatorFactoryBean.afterPropertiesSet(LocalValidatorFactoryBean.java:257) ~[spring-context-5.2.7.RELEASE.jar:5.2.7.RELEASE]
	at org.springframework.validation.beanvalidation.OptionalValidatorFactoryBean.afterPropertiesSet(OptionalValidatorFactoryBean.java:40) ~[spring-context-5.2.7.RELEASE.jar:5.2.7.RELEASE]
```

**Comment by al-niessner** [comment]
Fixed first error by `apt-get install libtcnative-1`. Still give me "cannot be reached" even though `mvn spring-boot:run` seems to be running. Looking into validation error now.

Still have dozens to hundreds of errors following stack traces like this one:

```
   WebServicesAutoConfiguration:
      Did not match:
         - @ConditionalOnClass did not find required class 'org.springframework.ws.transport.http.MessageDispatcherServlet' (OnClassCondition)
```

Without an example of what it should look like, cannot tell if this is good, indifferent, or bad.

**Comment by tloubrieu-jpl** [comment]
I have the 2 errors you showed in your previous post, but otherwise the server works.

**Comment by tloubrieu-jpl** [comment]
A example of working request is:
curl --location --request GET 'http://localhost:8080/products?limit=10'

I also have the 'did not match' messages.

The important thing for us is that you can request the server.

**Comment by al-niessner** [comment]
What if you point a browser at localhost:8080?

**Comment by tloubrieu-jpl** [comment]
I am getting the API UI

![Uploading image.png…]()

**Comment by al-niessner** [comment]
Here is my status:

last line of `mvn spring-boot:run` output:
`2021-02-25 21:18:26.411 DEBUG 94 --- [alina-utility-1] org.apache.catalina.session.ManagerBase  : End expire sessions StandardManager processingTime 2 expired sessions: 0`

Then result of curl:

```
$ curl --location --request GET 'http://localhost:8080/products?limit=10'
curl: (7) Failed to connect to localhost port 8080: Connection refused
```

which is an echo of my browser. application.properties server.port=8080 do not think it is wrong port. Just not sure why it is not responding other than it is not as started as it wants to be.

**Comment by al-niessner** [comment]
I think I am finding it. Looks like you specifically bind the socket to localhost rather than 0.0.0.0 letting the client localhost resolve as desired.

**Comment by tloubrieu-jpl** [comment]
@al-niessner I also have the DEBUG you shared with me, the server started line is slightly above it, right after the 'unconditional class' section, see:

`Unconditional classes:
----------------------

    org.springframework.boot.autoconfigure.context.ConfigurationPropertiesAutoConfiguration

    org.springframework.boot.autoconfigure.context.LifecycleAutoConfiguration

    org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration

    org.springframework.boot.autoconfigure.availability.ApplicationAvailabilityAutoConfiguration

    org.springframework.boot.autoconfigure.info.ProjectInfoAutoConfiguration



**2021-02-25 13:09:23.497  INFO 85157 --- [           main] g.n.pds.api.engineering.SpringBootMain   : Started SpringBootMain in 2.099 seconds (JVM running for 2.402)**
2021-02-25 13:10:22.579 DEBUG 85157 --- [alina-utility-1] org.apache.catalina.session.ManagerBase  : Start expire sessions StandardManager at 1614287422578 sessioncount 0
2021-02-25 13:10:22.580 DEBUG 85157 --- [alina-utility-1] org.apache.catalina.session.ManagerBase  : End expire sessions StandardManager processingTime 1 expired sessions: 0

`

**Comment by al-niessner** [comment]
Okay, it turns out that overriding the application.properties file in target/classes is not workable. When I change it then run `mvn spring-boot:run` mvn kindly puts the unworkable version back in place.

So, back to the original question of how can I override the properties file without having to edit the source? It does not appear as though I can pass arguments through mvn to the underlying application. While this might have worked for server.address because the spring framework seems to do some magic, it will not work for elasticSearch.host which is home grown although it does use `@Value` from spring so maybe it will. In the 10 minutes I scanned `@Value` it only appears to read the properties file. There are hints about how to pass a property file to spring, but that fails because mvn is in the way.

**Comment by al-niessner** [comment]
Mapping between the XML and [API docs](https://app.swaggerhub.com/apis/PDS_APIs/pds_federated_api/0.1.dev#/collections/get%20collection) is not 1:1. In the docs, there is 'title', 'ref', 'type', 'description':

![swagger_collection_result](https://user-images.githubusercontent.com/1130658/110524241-ff990e00-80c7-11eb-94c4-9c66c8b58e46.png)


The XML looks like:

```
<Investigation_Area>
  <name>MESSENGER</name>
  <type>Mission</type>
  <Internal_Reference>
    <lid_reference>urn:nasa:pds:context:investigation:mission.messenger</lid_reference>
    <reference_type>collection_to_investigation</reference_type>
  </Internal_Reference>
</Investigation_Area>
```

Is the mapping (XML -> API) something like:

```
name -> title
Internal_Reference.lid_reference -> ref
type -> type
?? -> description
```

At minimum what is description and a confirmation please that the rest are alright.

**Comment by tloubrieu-jpl** [comment]
Hi @al-niessner 
That is it. There is no description for investigation_area (from the model https://pds.nasa.gov/datastandards/documents/im/v1/index_1F00.html#10.21%C2%A0%C2%A0class_pds_investigation_area)

But there is one for the Observing System (https://pds.nasa.gov/datastandards/documents/im/v1/index_1F00.html#10.30%C2%A0%C2%A0class_pds_observing_system) and target identification (https://pds.nasa.gov/datastandards/documents/im/v1/index_1F00.html#10.39%C2%A0%C2%A0class_pds_target_identification) that we will work on later.

So for investigation area, the description is always null.

**Comment by tloubrieu-jpl** [comment]
Email thread with discussion on implementation of the external references in the API:Thanks for your feedback. So let’s do it this way s a first step:
{
    “investigations”: [ {
        “id”: “urn:nasa:pds:context:investigation:mission.orex”,
        “href”: “{baseUrl}/products/urn:nasa:pds:context:investigation:mission.orex”
    } ],
    observing_system_components: {
    [
        {
            “id”: “urn:nasa:pds:context:instrument_host:spacecraft.orex”,
            “href”: “{baseUrl}/products/urn:nasa:pds:context:instrument_host:spacecraft.orex”
        },
        {
            “id”: “urn:nasa:pds:context:instrument:ocams.orex”,
            “href”: {baseUrl}/products/urn:nasa:pds:context:instrument:ocams.orex”
        },
    ]
    },
 
And same for targets.
 
I left name,description,type in the specification but we will not populate them for now.
 
Al you can go on with the development: So you don’t need to read the XML as initially expected. But you will need to find the baseUrl from the springboot application to create the href.
 
I have updated the specification on swaggerhub (https://app.swaggerhub.com/apis/PDS_APIs/pds_federated_api/0.1.dev#/reference) and the model library (https://github.com/NASA-PDS/pds-api-javalib/tree/swaggerhub-pds_federated_api-0.1.dev).
 
Thanks,
 
Thomas
 
 
 
 
 
From: "Niessner, Albert F (US 398F)" <albert.f.niessner@jpl.nasa.gov>
Date: Wednesday, March 10, 2021 at 10:33 AM
To: "Loubrieu, Thomas G (US 398F)" <thomas.g.loubrieu@jpl.nasa.gov>, "Karpenko, Yevgen L (US 398F-Affiliate)" <yevgen.l.karpenko@jpl.nasa.gov>, "Padams, Jordan H (US 398A)" <jordan.h.padams@jpl.nasa.gov>
Cc: "Hughes, John S (US 398B)" <john.s.hughes@jpl.nasa.gov>
Subject: Re: PDSEN Breakout
 
 

My 2 cents, if it is worth even that, add href if desired by a set of end users but keep other information for scripts. When programmatically processing a query result, having only the choice of following yet another link is annoying at best. May only be needed in unit testing at the entropy end of the universe, but sending 100 extra bytes around is well worth it from my perspective.

 

--
Al Niessner
818.354.0859

--------
|  dS  | >= 0
--------
From: Loubrieu, Thomas G (US 398F)
Sent: Wednesday, March 10, 2021 9:32:05 AM
To: Karpenko, Yevgen L (US 398F-Affiliate); Padams, Jordan H (US 398A); Niessner, Albert F (US 398F)
Cc: Hughes, John S (US 398B)
Subject: Re: PDSEN Breakout
 
Hello,
 
Regarding that question, I propose the following plan:
 
Short term: have external references lid or lidvid (as in the pds4 label and registry doc)
Add name and type attributes to references (target, investigation_area, instruments)
Add a resolvable link that can be used by the client to “jump to the link resource”
 
The reason why I am delaying 2 and 3 are, consideration/questions, that I have:
 
2) If name and type comes from the context product, I feel like the API behavior would be more readable if we don’t bring attributes from other products (here context) into the requested product (for example collection). Besides I understand that the current name, type or description in  the investigation_area or instrument or target of a product are not well curated by the nodes, but I also feel like this could make sense to have specific values on the relationship between a product and for example its context target. I don’t have an good example but let’s try: Bennu is not the same to Osiris-rex as it would be to a spacecraft which would just pass by and take a picture (is that a realistic use case ?) and that type a relationship difference could be encoded in this name, type, description… attributes in the observational product or collection. In addition, one specific observational product or collection might need to describe specifically its targets relatively to each other (main, secondary…). I guess my questions are related to my lack of experience with the PDS4 model yet... My temporary conclusion is that it is better to ignore the relationship attributes completely from an API point view but I would not bring the context title or description in the Product description either (per the behavior readability motivation).
 
3) I like very much the idea of having resolvable URLs as external references (URIs), this is a best practice for restful APIs but I am annoyed by the fact the ids for pds are URN which are already supposed to be resolvable URIs as well although in an older fashion. So if we want to develop the idea of resolvable identifier (URI as URLs) in the API they would be for example http://pds.nasa.gov/api/products/urn:nasa:pds:context:investigation:mission.orex, which has some redundancy with the URN. So I was thinking we should stick to the lidvid URN until we have a better strategy to develop resolvable URI as URLs for PDS and the API. One possibility could be to have an canonical transformation from URN to API resolvable URLs. For example: ‘urn:nasa:pds:context:investigation:mission.orex’ would translated into ‘http://nasa.pds.gov/context/investigation/mission.orex’ that the API would resolve.
 
You can react to that now (if we read the email that far !) but I think that will be appropriate to have specific tickets for these 2 questions.
 
Thanks,
 
Thomas
 
 
 
 
From: "Karpenko, Yevgen L (US 398F-Affiliate)" <yevgen.l.karpenko@jpl.nasa.gov>
Date: Tuesday, March 9, 2021 at 3:13 PM
To: "Padams, Jordan H (US 398A)" <jordan.h.padams@jpl.nasa.gov>, "Arevalo, Eddie S (US 398H-Affiliate)" <eddie.arevalo@jpl.nasa.gov>, "Young, Jimmie D (US 398B)" <jimmie.d.young@jpl.nasa.gov>, "Suh, Catherine (US 398H-Affiliate)" <catherine.e.suh@jpl.nasa.gov>, "Kelly, Sean C (US 398G-Affiliate)" <sean.c.kelly@jpl.nasa.gov>, "Loubrieu, Thomas G (US 398F)" <thomas.g.loubrieu@jpl.nasa.gov>, "Wilson, Tishauna M (US 398F-Affiliate)" <tishauna.m.wilson@jpl.nasa.gov>, "Chau, Qui T (US 398F)" <qui.t.chau@jpl.nasa.gov>, "Barraza, Edward (US 397F)" <edward.barraza@jpl.nasa.gov>, "Collins, Scott (US 398F)" <scott.collins@jpl.nasa.gov>, "Niessner, Albert F (US 398F)" <albert.f.niessner@jpl.nasa.gov>
Cc: "tishauna2.wilson@gmail.com" <tishauna2.wilson@gmail.com>
Subject: RE: PDSEN Breakout
 
A comment about registry API and investigation / mission info.
 
Instead of displaying detailed mission info in a product API, just provide a (web) link to the mission resource, e.g.
 
{
    “lid”: “urn:nasa:pds:orex.ocams:data_calibrated:20190608t025825s015_map_radl2pan.fits”,
    “title”: “OSIRIS-REx OCAMS Level 2 (Calibrated) Radiance Science Image Data Product 2019-06-08T02:58:25.015Z”,
     …
    “Investigation”: {
        “lid”: “urn:nasa:pds:context:investigation:mission.orex”,
        “href”: “/api/context/urn:nasa:pds:context:investigation:mission.orex”
    },
    Observing_System_Component: {
    [
        {
            “lid”: “urn:nasa:pds:context:instrument_host:spacecraft.orex”,
            “href”: “/api/context/urn:nasa:pds:context:instrument_host:spacecraft.orex”
        },
        {
            “lid”: “urn:nasa:pds:context:instrument:ocams.orex”,
            “href”: “/api/context/urn:nasa:pds:context:instrument:ocams.orex”
        },
    ]
    },
 
}
 
 
Fancy keywords: HATEOAS (Hypermedia as the Engine of Application State), RFC 5988 (web linking)
 
 
Thanks,
Eugene

**Comment by tloubrieu-jpl** [comment]
Answer to Al's email:
"That seems opposite of what I was asking. Since rest is done can we just add href?"

Hi @al-niessner 

I understand the rest is done and that was indeed my specification in the ticket until yesterday but Jordan and Eugene said the information we are using in the PDS4 label is not reliable.

So we need to get it from somewhere else in the registry.

I would be ok, though if we keep that as an interim solution for this sprint and I can create a new ticket for the other implementation.
I don't think @jordanpadams @tdddblog would mind, would you ?

**Comment by al-niessner** [comment]
I need to be clear about this before deleting all of my previous work (extracting the investigation areas from PDS4 XML). I cut the text below from your previous post. Are you saying that these values are fixed no matter what and all I have to do is substitute the correct value for '{baseURL}'?

> {
“investigations”: [ {
“id”: “urn:nasa:pds:context:investigation:mission.orex”,
“href”: “{baseUrl}/products/urn:nasa:pds:context:investigation:mission.orex”
} ],
observing_system_components: {
[
{
“id”: “urn:nasa:pds:context:instrument_host:spacecraft.orex”,
“href”: “{baseUrl}/products/urn:nasa:pds:context:instrument_host:spacecraft.orex”
},
{
“id”: “urn:nasa:pds:context:instrument:ocams.orex”,
“href”: {baseUrl}/products/urn:nasa:pds:context:instrument:ocams.orex”
},
]
},

**Comment by tloubrieu-jpl** [comment]
Hi @al-niessner ,

You don’t need to delete your code. You should commit everything that you wrote to get information from the XML in your branch and push this branch on the repository https://github.com/NASA-PDS/registry-api-service.

If I followed that should do:
- get name/description/type/ref from XMLand add it to the product object. Optionally it could also add the href (but since it is a late request that does not need to).

You can commit and push that. That should close this ticket.

Then I will create a new ticket for what we discussed Tuesday and yesterday.

We will likely not use the code you wrote for this ticket but this will stay on github in case we need it in the future.

**Comment by al-niessner** [comment]
@tloubrieu-jpl

All code on branch is up to date and working (nothing removed). Moving onto #52
